Extended Blocks: add Loom as oEmbed provider

retrofox opened the pull request:
This PR adds loom.com as a new oembed provider, allowing embed resources from this service.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Add loom.com as a new oembed provider

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Edit/Add a post
* Confirm you see the `Loom` variation among block options.

<img width="500" alt="Screen Shot 2020-09-10 at 9 28 43 PM" src="https://user-images.githubusercontent.com/77539/92831384-b849f300-f3ac-11ea-85ed-684cd52e8e9d.png">

* Add Loom embed variation to the editor. You can use this link https://www.loom.com/share/e3dcec661c37487b818b8e3b8225ec27

<img src="https://user-images.githubusercontent.com/77539/92831735-2098d480-f3ad-11ea-82a0-b8055a0e13c5.png" width="500px" />

* Finally, confirm the resource is properly rendered.

![image](https://user-images.githubusercontent.com/77539/92801481-0bf81480-f38c-11ea-83df-383ab03f4675.png)

* Confirm, you can paste the resource and it will be automatically transformed in a `core/embed` block, `Loom` variation.

Confirm it works with WordPress greater and lower than 5.5

<img width="400" src="https://user-images.githubusercontent.com/77539/94067274-035c0100-fdc4-11ea-998a-43702537e9d6.png" />



#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add loom.com as a new oembed provider

creativecoder reviewed (commented):
Note, I tested with different versions...

**WP 5.5 plus latest Gutenberg plugin** - works as expected

**WP 5.5** - works as expected

**WP 5.4 (previous version)** - `registerBlockVariation` is available, but the embed block did not use variations in this version, so the block isn't available. But the embed still works through the embed block, or by pasting a loom link.

creativecoder commented:
I'm having trouble getting the embed to preview within the editor. I works fine on the front end of the site within the post/page, though.

Seems to be related to the `sandbox` attribute on the iframe. The iframe embed works fine pasted into an html block when I remove that attribute.

<img width="1289" alt="Screen Shot 2020-09-10 at 17 58 33" src="https://user-images.githubusercontent.com/1699996/92821485-bb36ea80-f390-11ea-8ba3-f64191c16d2c.png">

retrofox commented:
> Note, I tested with different versions...
> 
> **WP 5.5 plus latest Gutenberg plugin** - works as expected
> 
> **WP 5.5** - works as expected
> 
> **WP 5.4 (previous version)** - `registerBlockVariation` is available, but the embed block did not use variations in this version, so the block isn't available. But the embed still works through the embed block, or by pasting a loom link.

I've [added a fallback](60624b1271d5b50eea2449362dff032486159c88) for versions lower than 5.5, using the `editor.registerBlockType` filter. It works very well for me. it would be nice if you can check it.

retrofox commented:
> Seems to be related to the `sandbox` attribute on the iframe. The iframe embed works fine pasted into an html block when I remove that attribute.

Can you provide the Loom resource (URL), please? I think it was a matter of your testing flow. Keep in mind that the HTML markup is cached and if it was not properly generated in the first attempt it will keep returning the wrong code beyond the app adds the Loop provider properly.
I say this because we are seeing the `sandbox` attribute. It shouldn't appear once the provider is added.

creativecoder commented:
> Keep in mind that the HTML markup is cached and if it was not properly generated in the first attempt it will keep returning the wrong code beyond the app adds the Loop provider properly.

I think that was it. My issues were resolved when I switched to a different test site. The loom url I'm using for testing is https://www.loom.com/share/0281766fa2d04bb788eaf19e65135184

jeherve commented:
Do you happen to have some background on this addition?

- Shouldn't this be added to Core directly instead?
- In the past, we've added embeds without dealing with the block editor UI.
https://github.com/Automattic/jetpack/blob/f438233624c074bf4cc54755d7397d706cb801d3/modules/shortcodes/others.php#L1-L10
(also see D49210-code for where this is managed on WordPress.com)

My first thought would be that the embed addition would be a good fit for Jetpack / WordPress.com, without a UI, until the embed makes it way into Core.

retrofox commented:
> > Keep in mind that the HTML markup is cached and if it was not properly generated in the first attempt it will keep returning the wrong code beyond the app adds the Loop provider properly.
> 
> I think that was it. My issues were resolved when I switched to a different test site. The loom url I'm using for testing is https://www.loom.com/share/0281766fa2d04bb788eaf19e65135184

It works for me. 👍 

![image](https://user-images.githubusercontent.com/77539/92921808-1efdf980-f40b-11ea-9636-0a8f62da5ca2.png)

retrofox commented:
Thanks, Jeremy.

> * Shouldn't this be added to Core directly instead?
We discussed a little bit about this and concluded adding this provider to Jetpack could be the whole implementation faster. It doesn't mean that we couldn't start adding Loom straight in the core but maybe in the following step?
We'd really like to implement Loom as soon as possible, and since it doesn't seem to have technical blockers for its implementation, Jetpack was the best candidate.

> * In the past, we've added embeds without dealing with the block editor UI.

Well... maybe in the past the block editor didn't exist :-D.
I put the backend code in `./extensions/extended-blocks/core-embed` because, basically, we are extending the core/embed block. 
Something similar happens with `core/video`, `core/cover` and `core/audio`. Although the extending reasons are different, in the end, we modifying blocks that are not registered by Jetpack.

>   https://github.com/Automattic/jetpack/blob/f438233624c074bf4cc54755d7397d706cb801d3/modules/shortcodes/others.php#L1-L10

I didn't want to edit this file because it belongs to `shortcodes`. It isn't a big deal moving changes straight there.

>   (also see D49210-code for where this is managed on WordPress.com)

Yeah, but it's for Simple sites, correct? In that case, we'll need to add the provider to two different files?

> My first thought would be that the embed addition would be a good fit for Jetpack / WordPress.com, without a UI, until the embed makes it way into Core.

That was our first thought too, but then we changed in order to reduce the landing time. Again, we can start to work on moving the whole implementation to the core. It means, if I'm not wrong, add the provider to WordPress core and also add the block variation in Gutenberg core. Is it correct?

> If we decide to add a UI in the editor, maybe we should do it for all our other embeds as well?

Do you mean to add a block variation for all oEmbed blocks in Jetpack? If so, I'd happy to take over this task.

jeherve commented:
Related discussion: 
p1599826045429400-slack-CKZHG0QCR

> adding this provider to Jetpack could be the whole implementation faster. It doesn't mean that we couldn't start adding Loom straight in the core but maybe in the following step?

👍 

> it's for Simple sites, correct? In that case, we'll need to add the provider to two different files?

Correct, unless we were to bring files in sync in the future. Here was an exploration:

D27182-code

> It means, if I'm not wrong, add the provider to WordPress core and also add the block variation in Gutenberg core. Is it correct?

Exactly. Adding an embed to Core is usually done in two patches these days: one to add it to Gutenberg, one to add it to Core. Here is an example:
- 49083-core
- https://github.com/WordPress/gutenberg/pull/19345/

> Do you mean to add a block variation for all oEmbed blocks in Jetpack? If so, I'd happy to take over this task.

Yes, exactly. If we think it's a valuable addition for the Loom provider, and if you've developed the basic structure for one provider, we may as well add more in the future.

jeherve commented on extensions/extended-blocks/core-embed/core-embed.php:
```diff
@@ -0,0 +1,4 @@
+<?php
+
+// Add Loom.com as an oEmbed provider.
+wp_oembed_add_provider( '#https?://(www\.)?loom\.com/share/.*#i', 'https://www.loom.com/v1/oembed', true );
```
Do you think you could add a simple test for this, kind of like what we have here?
https://github.com/Automattic/jetpack/blob/master/tests/php/modules/shortcodes/test-class.others.php

jeherve commented on extensions/extended-blocks/core-embed/variations.js:
```diff
@@ -0,0 +1,48 @@
+const embedLoomIcon = {
+	foreground: getIconColor(),
+	src: (
+		<SVG viewBox="0 0 48 48">
```
Should we move this icon to `extensions/shared/icons.js`?

jeherve commented on tests/php/modules/shortcodes/test-class.loom.php:
```diff
@@ -0,0 +1,69 @@
+<?php
```
Since your provider is declared in `modules/shortcodes/others.php`, I think you could move that test to `tests/php/modules/shortcodes/test-class.others.php` with the rest of the "other" tests.

retrofox replied:
✅

jeherve reviewed (commented):
This tests well. I only have one suggestion.

jeherve commented:
r214751-wpcom
